lib/wallet/rpc.js

```diff
@@ -1049,10 +1049,12 @@ class RPC extends RPCBase {
       const entry = await this.client.getEntry(block);
       if (entry)
         height = entry.height;
+      else
+        throw new RPCError(errs.MISC_ERROR, 'Block not found');
     }
 
     if (height === -1)
-      height = this.chain.height;
+      height = chainHeight;
```

Comment:
I don't think we need this if statement at all. When no blockhash is passed, we want to return all transactions. Since the check is if `(wtx.height < height)`, if will add all transactions to the result, when height is -1.

Also, comparing against the core, it seems to exclude the current block being passed (e.g. when latest block is passed, it doesn't return any confirmed transactions). So the above check should really be `if (wtx.height <= height)`.

Reply:
So, more like what I had back here?

https://github.com/bcoin-org/bcoin/pull/537#issuecomment-406367869

...but with a `<=` like you say

Reply:
Maybe you meant `if (wtx.height <= height  && wtx.height !== -1)`? In which case it would list unconfirmed transactions, as core does, even if `lastblock` is omitted.

Reply:
Here's that snippet again.

So if no block is passed, `height = -1` and stays `-1`. Then when it iterates the `wtx of txs`, it will process and return every `wtx` because of `&& height !== -1`. 

If an invalid block is passed, an error is thrown and process stops (like Core https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.15.1.md#low-level-rpc-changes)

If a valid block is passed, `height` WILL be changed, to height of passed block. Then the `wtx of txs` will skip (`continue`) any `wtx` at that `height` and below.

This would mean we don't need the current height of the blockchain (https://github.com/bcoin-org/bcoin/issues/536) and could even remove the line `const chainHeight = this.wdb.state.height;` at the top of the function.

```
    let height = -1;

    if (block) {
      const entry = await this.client.getEntry(block);
      if (entry)
        height = entry.height;
      else
        throw new RPCError(errs.MISC_ERROR, 'Block not found');
    }

    const txs = await wallet.getHistory();
    const out = [];

    let highest = null;

    for (const wtx of txs) {
      if (wtx.height <= height && height !== -1)
        continue;
    
      <... add wtx to return JSON>
    }
```